src/System.Private.CoreLib/shared/System/Exception.cs

```diff
@@ -164,6 +166,14 @@ public override string ToString()
             return s;
         }
 
+        protected internal string InnerExceptionToString()
```

Comment:
I am not sure whether avoiding a bit of the code duplication at the cost of introducing extra allocation is an improvement.

Reply:
Also, keeping the code inlined like it used to be makes it easier to see what the output is going to look like.

Reply:
Maybe just introduce a constant for `" ---> "`?

Reply:
The real reason to inline is to centralize the indenting in the followup change. I could remove it if you like just for this change.

BTW it's only an extra allocation for `COMException` here isn't it? The others all concatenate strings anyway. (Although, I could change those, but in that case I could also make this return a StringBuilder, for them to do `sb.Append(otherSb)`)

I notice in the followup if I don't extract this I will also need to remove the divider line from XslException.

Reply:
Looking at the types that need this (because they want to format the InnerException themselves) it's because they want to put something between the message and the stack. Eg., COMException puts the HRESULT, FileXXException puts the filename etc. Possibly they should include these in the message, then they might not need to override ToString() at all.

Reply:
Yes, I think it would be better to not future proof this.

It is an extra allocation for most cases, e.g. the BadImageFormatException is going to also have an extra allocation.

> XslException

XslException does not live in CoreLib, so this sounds like the follow up change will be introducing a new public APIs...

If the follow up change is going to affect XslException, then you would

Reply:
> XslException does not live in CoreLib, so this sounds like the follow up change will be introducing a new public APIs...

You're right, I'll avoid a public API...

> It is an extra allocation for most cases, e.g. the BadImageFormatException is going to also have an extra allocation.

I'm missing something - BIFE was previously concatenating strings, now it is still concatenating strings (albeit plus a newline, which is the point of the change)

But not a problem -- I"ll inline this again...

Reply:
Before your change: 

`s = s + " ---> " + InnerException.ToString();` compiles into `string.Concat(s, " ---> ",  InnerException.ToString())`

After your change:

```
string temp = string.Concat(Environment.NewLine, " ---> ", _innerException.ToString()); // this is the extra allocation
string.Concat(s, temp);
```

Reply:
oh! I forgot that `+=` is compiled into `Concat`. thank you